src/proxy/resolve.rs

```diff
@@ -107,61 +125,442 @@ where
+            Ok(Async::Ready(ep))
+        }
+    }
+
+    #[derive(Debug)]
+    struct Svc<T>(Vec<oneshot::Receiver<T>>);
+    impl<T> Service<()> for Svc<T> {
+        type Response = T;
+        type Error = oneshot::error::RecvError;
+        type Future = oneshot::Receiver<T>;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            Ok(().into())
+        }
+
+        fn call(&mut self, _: ()) -> Self::Future {
+            self.0.pop().expect("exhausted")
+        }
+    }
+
+    #[test]
+    fn inserts_delivered_out_of_order() {
+        with_task(move || {
+            let (mut reso_tx, resolution) = mpsc::channel(2);
+            let (make0_tx, make0_rx) = oneshot::channel::<Svc<usize>>();
+            let (make1_tx, make1_rx) = oneshot::channel::<Svc<usize>>();
+            let make = Svc(vec![make1_rx, make0_rx]);
+
+            let mut discover = Discover::new(resolution, make);
+            assert!(
+                discover.poll().expect("discover can't fail").is_not_ready(),
+                "ready without updates"
+            );
+
+            let addr0 = SocketAddr::from(([127, 0, 0, 1], 80));
+            reso_tx.try_send(Update::Add(addr0, ())).unwrap();
+            assert!(
+                discover.poll().expect("discover can't fail").is_not_ready(),
+                "ready without service being made"
+            );
+            assert_eq!(
+                discover.makes.futures.len(),
+                1,
+                "must be only one pending make"
+            );
+            assert_eq!(
+                discover.makes.cancelations.len(),
+                1,
+                "no pending cancelation"
+            );
+
+            let addr1 = SocketAddr::from(([127, 0, 0, 2], 80));
+            reso_tx
+                .try_send(Update::Add(addr1, ()))
+                .expect("update must be sent");
+            assert!(
+                discover.poll().expect("discover can't fail").is_not_ready(),
+                "ready without service being made"
+            );
+            assert_eq!(
+                discover.makes.futures.len(),
+                2,
+                "must be only one pending make"
+            );
+            assert_eq!(
+                discover.makes.cancelations.len(),
+                2,
+                "no pending cancelation"
+            );
+
+            let (rsp1_tx, rsp1_rx) = oneshot::channel();
+            make1_tx
+                .send(Svc(vec![rsp1_rx]))
+                .expect("make must receive service");
+            match discover.poll().expect("discover can't fail") {
+                Async::NotReady => panic!("not processed"),
+                Async::Ready(Change::Remove(..)) => panic!("unexpected remove"),
+                Async::Ready(Change::Insert(a, mut svc)) => {
+                    assert_eq!(a, addr1);
+
+                    assert!(svc.poll_ready().unwrap().is_ready());
+                    let mut fut = svc.call(());
```

Comment:
Checking my understanding here: Would this `svc.call(())` pop the last element out of the Vec -- which is `make0_rx`? I'm looking at how `fn call` is implicted for `Svc`, and see that we just pop last element off of the Vec of services. It was put in though as `make1_rx` and then `make0_rx`

Reply:
The discover.poll() returns the service from `make1_rx1`... Then, this `call` returns `rsp1_rx`.